### ISSUE TITLE
Resolve method is not being called in a nested schema

### ISSUE THREAD

**Issue by millejon** [vitalik/django-ninja#1222]
Hey everyone, I am trying to output a nested schema in a response but the resolve method of the nested schema is not being called and therefore, the following error is being raised: 

The main schema for the response (shortened for brevity):

```
class DetailedAlbumOut(Schema):
    id: int
    artists: List[ArtistOut]
    title: str
```

The ArtistOut schema used:
```
class ArtistOut(Schema):
    id: int
    name: str
    url: str

    @staticmethod
    def resolve_url(obj):
        artist_url = reverse("api-1.0:retrieve_artist", kwargs={"id": obj.id})
        return obj.request.build_absolute_uri(artist_url)
```

The artist object only has id and name fields. The URL attribute is calculated using the resolve method. When I use the ArtistOut schema on its own, it works perfectly. However, when I try to output the DetailedAlbumOut response, I get the following error: 

```
pydantic_core._pydantic_core.ValidationError: 1 validation error for NinjaResponseSchema
response.artists.0.url
  Field required [type=missing, input_value=<DjangoGetter: <Artist: Jay-Z>>, input_type=DjangoGetter]
```

From playing around with it to try to figure out how to get it to work, it seems like the resolve_url method is not being called. If I remove the url field and the resolve method, then the output works in that I get the id and name of each artist, but I would like to have the the url to the artist resource included.

Any help is much appreciated. Let me know if you have any further questions or need anything to be clarified, thank you!

**Comment by vitalik** [comment]
Hi @millejon 

could you please give some reproducable example (with view or something ) so I can investigate

**Comment by millejon** [comment]
Hey @vitalik! I'm not sure how much you need from my program to properly investigate so I'm giving you quite a bit of it, let me know if you need anything more.

Here are the schema used:
```
class ArtistIn(Schema):
    name: str
```
```
class ArtistOut(Schema):
    id: int
    name: str
    url: str

    @staticmethod
    def resolve_url(obj):
        artist_url = reverse("api-1.0:retrieve_artist", kwargs={"id": obj.id})
        return obj.request.build_absolute_uri(artist_url)
```
```
class AlbumIn(Schema):
    artists: List[int]
    title: str
    release_date: date
```
```
class AlbumOut(Schema):
    id: int
    artists: List[ArtistOut]
    title: str
    release_date: date
    url: str

    @staticmethod
    def resolve_url(obj):
        album_url = reverse("api-1.0:retrieve_album", kwargs={"id": obj.id})
        return obj.request.build_absolute_uri(album_url)
```

You should be able to reproduce the error when creating an album. The endpoints for creating an artist and an album are:
```
@api.post("artists/", response={201: schema.ArtistOut}, tags=["artists"])
def create_artist(request, data: schema.ArtistIn):
    artist = models.Artist.objects.create(**data)
    artist.request = request
    return 201, artist
```
```
@api.post("albums/", response={201: schema.AlbumOut}, tags=["albums"])
def create_album(request, data: schema.AlbumIn):
    artists = data.pop("artists")

    album = models.Album.objects.create(**data)
    for artist in artists:
        album.artists.add(artist)
    album.save()

    album.request = request
    return 201, album
```
The error I am getting is:
```
pydantic_core._pydantic_core.ValidationError: 1 validation error for NinjaResponseSchema
response.artists.0.url
  Field required [type=missing, input_value=<DjangoGetter: <Artist: Jay-Z>>, input_type=DjangoGetter]
    For further information visit https://errors.pydantic.dev/2.7/v/missing
```

**Comment by vitalik** [comment]
@millejon you never mentioned django ninja version you are using...

but here is my test on latest (1.2.1) - I mocked models :

mocks:

```python

class Objects:
    def __init__(self, model):
        self.model = model

    def create(self, **kwargs):
        return self.model(**kwargs)


class Artist:
    def __init__(self, name, id=1):
        self.id = id
        self.name = name


Artist.objects = Objects(Artist)


class M2MList(list):
    def add(self, id):
        self.append(Artist('John', id=id))


class Album:
    def __init__(self, title, release_date):
        self.id = 1
        self.title = title
        self.release_date = release_date
        self.artists = M2MList()

    def save(self):
        pass


Album.objects = Objects(Album)
```

Code:

```Python
class ArtistIn(Schema):
    name: str


class ArtistOut(Schema):
    id: int
    name: str
    url: str

    @staticmethod
    def resolve_url(obj, context):
        artist_url = f"/artists/{obj.id}/"
        # artist_url = reverse("api-1.0:retrieve_album", kwargs={"id": obj.id})
        return context["request"].build_absolute_uri(artist_url)


class AlbumIn(Schema):
    artists: List[int]
    title: str
    release_date: date


class AlbumOut(Schema):
    id: int
    artists: List[ArtistOut]
    title: str
    release_date: date
    url: str

    @staticmethod
    def resolve_url(obj, context):
        # album_url = reverse("api-1.0:retrieve_album", kwargs={"id": obj.id})
        album_url = f"/albums/{obj.id}/"
        return context["request"].build_absolute_uri(album_url)


@api.post("artists/", response={201: ArtistOut}, tags=["artists"])
def create_artist(request, payload: ArtistIn):
    data = payload.dict()
    artist = Artist.objects.create(**data)
    return 201, artist


@api.post("albums/", response={201: AlbumOut}, tags=["albums"])
def create_album(request, payload: AlbumIn):
    data = payload.dict()
    artists = data.pop("artists")

    album = Album.objects.create(**data)
    for artist in artists:
        album.artists.add(artist)
    album.save()

    album.request = request
    return 201, album
 ```


Some notes

 - if you add `context` to your resolver - you can get request from context (so you do not have to use this trick with `album.request = request` )
 -  your code was not working righ away - so I had to extract dicts from schemas `data = payload.dict()` 
 - reverse I commented out as I would need to to create all these urls (maybe you should try that too - as this might be the issue)

overall it works for me just fine: 

<img width="788" alt="SCR-20240709-kqpv" src="https://github.com/vitalik/django-ninja/assets/95222/91010af6-79ee-4c4d-87cf-a81104bbae0d">

**Comment by millejon** [comment]
@vitalik Thank you for putting so much effort into looking into this for me!

My bad about the extracting dictionary part. I took the line out of my example where I did this because I call a utility function to strip whitespace from the user input, but I forgot to put the dictionary extraction back in there.

After working backwards from your working example, I am pretty sure the problem was because of that ```album.request = request``` trick I was doing in each endpoint. When I removed that and put the context in the resolver like you suggested, it worked fine, even when leaving in the reverse function calls.

Thank you again for helping me figure this out!
